server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -522,150 +529,234 @@ public IndexMetaDataGenerations indexMetaDataGenerations() {
-                                }
-                            } else if (SHARD_GENERATIONS.equals(indexMetaFieldName)) {
-                                XContentParserUtils.ensureExpectedToken(
-                                    XContentParser.Token.START_ARRAY, parser.currentToken(), parser::getTokenLocation);
-                                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
-                                    gens.add(parser.textOrNull());
-                                }
-
-                            }
-                        }
-                        assert indexId != null;
-                        indexSnapshots.put(indexId, Collections.unmodifiableList(snapshotIds));
-                        for (int i = 0; i < gens.size(); i++) {
-                            String parsedGen = gens.get(i);
-                            if (fixBrokenShardGens) {
-                                parsedGen = ShardGenerations.fixShardGeneration(parsedGen);
-                            }
-                            if (parsedGen != null) {
-                                shardGenerations.put(indexId, i, parsedGen);
-                            }
-                        }
-                    }
-                } else if (INDEX_METADATA_IDENTIFIERS.equals(field)) {
+        Map<String, String> indexMetaIdentifiers = null;
+        while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+            final String field = parser.currentName();
+            switch (field) {
+                case SNAPSHOTS:
+                    parseSnapshots(parser, snapshots, snapshotStates, snapshotVersions, indexMetaLookup);
+                    break;
+                case INDICES:
+                    parseIndices(parser, fixBrokenShardGens, snapshots, indexSnapshots, indexLookup, shardGenerations);
+                    break;
+                case INDEX_METADATA_IDENTIFIERS:
                     if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
                         throw new ElasticsearchParseException("start object expected [" + INDEX_METADATA_IDENTIFIERS + "]");
                     }
-                    indexMetaIdentifiers.putAll(parser.mapStrings());
-                } else if (MIN_VERSION.equals(field)) {
+                    indexMetaIdentifiers = parser.mapStrings();
+                    break;
+                case MIN_VERSION:
                     if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
                         throw new ElasticsearchParseException("version string expected [min_version]");
                     }
                     final Version version = Version.fromString(parser.text());
                     assert SnapshotsService.useShardGenerations(version);
-                } else {
-                    throw new ElasticsearchParseException("unknown field name  [" + field + "]");
-                }
+                    break;
+                default:
```

Comment:
default clause is not part of the linked patch, had to add it because of checkstyle